File: javascript/protein-translation/protein-translation.js
const CODONS = new Map([
    ['AUG', 'Methionine'],
    ['UUU', 'Phenylalanine'],
    ['UUC', 'Phenylalanine'],
    ['UUA', 'Leucine'],
    ['UUG', 'Leucine'],
    ['UCU', 'Serine'],
    ['UCC', 'Serine'],
    ['UCA', 'Serine'],
    ['UCG', 'Serine'],
    ['UAU', 'Tyrosine'],
    ['UAC', 'Tyrosine'],
    ['UGU', 'Cysteine'],
    ['UGC', 'Cysteine'],
    ['UGG', 'Tryptophan'],
    ['UAA', 'STOP'],
    ['UAG', 'STOP'],
    ['UGA', 'STOP']
]);

export function translate(codons = '') {
    let proteins = [];
    let left = codons;
    while (left.length != 0) {
        let codon = left.slice(0, 3);
        left = left.slice(3);
        let protein = CODONS.get(codon);
        if (protein == undefined) {
            throw new Error('Invalid codon');
        }
        if (protein == 'STOP') {
            break;
        }
        proteins.push(protein);
    }
    return proteins;
}

File: clojure/protein-translation/src/protein_translation.clj
(ns protein-translation
  (:require [clojure.string :as string]))

(def ^:private proteins-to-codons
  {"Methionine"    ["AUG"]
   "Phenylalanine" ["UUU" "UUC"]
   "Leucine"       ["UUA" "UUG"]
   "Serine"        ["UCU" "UCC" "UCA" "UCG"]
   "Tyrosine"      ["UAU" "UAC"]
   "Cysteine"      ["UGU" "UGC"]
   "Tryptophan"    ["UGG"]
   "STOP"          ["UAA" "UAG" "UGA"]})

(def ^:private codons-to-protein
  (into {} (for [[protein codons] proteins-to-codons
                 codon codons]
             [codon protein])))

(defn translate-codon [codon]
  (codons-to-protein codon))

(defn translate-rna [rna]
  (->> rna
       (partition 3)
       (map string/join)
       (map translate-codon)
       (take-while (partial not= "STOP"))))
